src/rt/cpp/token.h

```diff
@@ -0,0 +1,305 @@
+// Copyright Microsoft and Project Verona Contributors.
+// SPDX-License-Identifier: MIT
+#pragma once
+
+#include "cown.h"
+#include "promise.h"
+#include "when.h"
+
+#include <atomic>
+#include <cstdint>
+
+namespace verona::cpp
+{
+  /**
+   * An approach to limiting the work in the system.
+   *
+   * It provides a bounded set of Tokens, that can be requested
+   * from a Token::Source.  If the there are no tokens available,
+   * then the lambda is delayed, until a Token has been destroyed.
+   *
+   * Tokens must be treated linearly, once a token is no longer required
+   * its destructor runs, and that enables the source to provide another token.
+   */
+  class Token
+  {
+    /**
+     * Internal function that provides a cown that can be waited on.
+     * This is like a create promise, just the minimal thing required for this
+     * example.
+     */
+    static cown_ptr<size_t> make_wait()
```

Comment:
How about using verona::rt::Promise for that? It might be easier to follow.

Reply:
Yeah, I would like to tidy up the Promise code a little to make it fit nicely here, but then I was going to make a larger change.  I'll move to promises before we integrate this, if we decide too.